Issue #10140: Introduce GeckoStorageDelegateWrapper to replace GeckoLoginDelegateWrapper

gabrielluong opened the pull request:
Fixes #10140 

Reviewing individual commits might make this easier.

Part 1: Refactor autocomplete extension functions into ext/Login.kt
- Moves `Autocomplete.LoginEntry.toLogin()` and `Login.toLoginEntry()` extension functions
into `ext/Login.kt`
- Remove duplicates of `Autocomplete.LoginEntry.toLogin()` and `Login.toLoginEntry()`

Part 2: Make a copy of GeckoLoginDelegateWrapper as GeckoStorageDelegateWrapper
- We make a copy of GeckoLoginDelegateWrapper to not break the deprecated usage of `Autocomplete.LoginStorageDelegate` until we are ready to replace with `GeckoStorageDelegateWrapper` which will use the new GV API `Autocomplete.StorageDelegate`.

Part 3: Implement the new Autocomplete.StorageDelegate interface in GeckoStorageDelegateWrapper
- Uses the new GV `Autocomplete.StorageDelegate` interface in GeckoStorageDelegateWrapper
- Implements GeckoView's `Autocomplete.StorageDelegate.onCreditCardFetch`

------------------------------------------------------

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

grigoryk commented on components/concept/storage/build.gradle:
```diff
@@ -27,6 +27,8 @@ dependencies {
     // Included via 'api' because this module is unusable without coroutines.
     api Dependencies.kotlin_coroutines
+
+    api project(':lib-dataprotect')
```
Let's instead move over `KeyProvider` and `ManagedKey` into `concept-storage`. None of our concepts depend on any other modules, which is a nice pattern I'd like to keep.

These two interfaces now become part of the storage definitions, which I think makes sense here. E.g. to interact with a storage, you need to be aware of these key interfaces.

And, libdataprotect doesn't actually provide a concrete implementations for these interaces - that happens as part of our storage impl for autofill stuff. So that aligns pretty well as well - concept-storage interfaces implemented by a specific storage module.

grigoryk reviewed (changes requested):
This looks good to me, except for the dependency on lib-dataprotect which we can easily avoid here.

gabrielluong commented:
@grigoryk Since your last review, I renamed `GeckoStorageDelegateWrapper` to `GeckoAutocompleteStorageDelegate`. I dropped the "Wrapper" and reused the existing naming convention I saw that simply added a `Gecko` prefix to all the GV delegates.

Also, added a Part 4 which adds test for GeckoAutocompleteStorageDelegate.

gabrielluong commented:
> @grigoryk Since your last review, I renamed `GeckoStorageDelegateWrapper` to `GeckoAutocompleteStorageDelegate`. I dropped the "Wrapper" and reused the existing naming convention I saw that simply added a `Gecko` prefix to all the GV delegates.
> 
> Also, added a Part 4 which adds test for GeckoAutocompleteStorageDelegate. This changes our storage delegates returns of `Deferred`s into a `suspend fun` to simplify testing. This was already planned feedback from our other PR regarding credit card storage validation.

I am gonna move Part 4 into a separate PR since I think I need some help to make the tests pass.

Amejia481 commented on components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/autofill/GeckoAutocompleteStorageDelegate.kt:
```diff
@@ -0,0 +1,81 @@
+    override fun onCreditCardFetch(): GeckoResult<Array<Autocomplete.CreditCard>>? {
+        val result = GeckoResult<Array<Autocomplete.CreditCard>>()
+
+        GlobalScope.launch(IO) {
```
You could have the dispatcher as internal variable `@VisibleForTesting` and change it from tests, that will make testing reliable and simple.  Same for other usages

Amejia481 reviewed (commented):
I looks good to me, there multiple sections that are not covered by test on `GeckoAutocompleteStorageDelegate` could we add test to covered these sections?

like
<img width="1388" alt="image" src="https://user-images.githubusercontent.com/773158/117199872-b825ba00-adb8-11eb-9775-b15c95874247.png">

grigoryk commented:
Thanks @gabrielluong!
